### PR TITLE
Add prometheus installer with persistent volumes to perf

### PR DESCRIPTION
**Pull request by therealmitchconnors**
This enables us to keep more consistent records regarding perf test results.

### CODE REVIEW

**Comment by howardjohn** [comment]
I am looking into adding prometheus alarms for the perf cluster as well, so this will likely be useful since we would need to customize the prometheus install for that as well

**Comment by therealmitchconnors** [comment]
> I am looking into adding prometheus alarms for the perf cluster as well, so this will likely be useful since we would need to customize the prometheus install for that as well

I have disabled alerts in the prometheus-operator-values.yaml, but it can easily be re-enabled, allowing you to manipulate alerts directly with kubectl.  See https://github.com/coreos/prometheus-operator/blob/master/Documentation/user-guides/alerting.md

**Comment by howardjohn** [comment]
> > I am looking into adding prometheus alarms for the perf cluster as well, so this will likely be useful since we would need to customize the prometheus install for that as well
> 
> I have disabled alerts in the prometheus-operator-values.yaml, but it can easily be re-enabled, allowing you to manipulate alerts directly with kubectl. See https://github.com/coreos/prometheus-operator/blob/master/Documentation/user-guides/alerting.md

Perfect, thanks for the info

**Comment by costinm** [comment]
Note that this change means we'll no longer test the istio that we ship to customers...

We should ship/recommend this to customers - but it's not there yet. 
Can we apply this to our official templates, or change the default installer to use this new style ? 
@douglas-reid

**Comment by douglas-reid** [comment]
Can we have some background here?  Why not support, for all use cases, the option to install with the prometheus operator?

I think that this change should come with an associated issue describing exactly what is being desired and how this is solving that need.

I'd also like to understand why we are duplicating scrape config and why prometheus is being disabled by default here.

FWIW, there are already existing issues / PRs that may be relevant here:
- https://github.com/istio/istio/issues/10134
- https://github.com/istio/istio/pull/9508

Perhaps this should be a discussion for the next P&T WG.

**Comment by therealmitchconnors** [comment]
@douglas-reid, the PRs you referenced appear to cover precisely what @costinm asked, namely, merging these sorts of changes into the primary istio chart.  I expect that, once those issues are handled, we should be able to discard many of these changes and use the default chart.  

In the mean time, we need our qual tests to stop discarding perf data by storing prometheus information in ephemeral drives.  This change will enable us to collect better information regarding our release candidates without waiting for the primary chart to be updated.

**Comment by therealmitchconnors** [comment]
@costinm, I was unaware of your efforts.  I suggest that we move the prometheus deployment to the new namespace once your installer is adopted by this test suite.  There will be multiple other changes necessary at that time, and (as @douglas-reid commented), we may no longer need specific qual test prometheus config by then...

**Comment by douglas-reid** [comment]
@therealmitchconnors I understand what you are saying, but worry about the approach. I'm still not sure I've seen any real qualification of the issue at hand here, outside of "This enables us to keep more consistent records regarding perf test results."

Why can we not just do the work up front to fully support Prometheus operator-based installations alongside the traditional installs?  Why not take an alternate approach?

In the short-term, why not, for instance, use `remote-write` configuration to write to a permanent storage location, or use Prometheus federation? Or use something like Thanos with GCS backup here?

This changes adds a bunch of stuff beyond just persistent disk to prometheus. As far as I can tell, it adds kube-state-metrics, etc.

There are many ways to productionizing Prometheus. I'd like to understand which ones we are going to support in the appropriate WG and not just push this through.

**Comment by mandarjog** [comment]
@therealmitchconnors do we need to install a separate scrape config here, can't we use the one that is shipped with the release?

@douglas-reid This is adding to the tools and perf cluster. It may not be *the* way to productionize prometheus. It will however satisfy our qual cluster requirements.
If we feel that supporting a productionized prometheus / operator based install is part of istio proper, we can have that discussion in the WG. I can think of this as the testing ground for that work.


Here are my requirements for prometheus in the qual cluster
1. Be reliable, do not die
2. Store prometheus data for several days to weeks.
@jnativio  already runs with a persist prometheus instance. 
Can you share those details here?

**Comment by therealmitchconnors** [comment]
@mandarjog, the scrape config is copied verbatim from the istio chart, but with several lines removed.  While we could do this line removal automatically, my bash fu is not that strong.  I welcome any pull requests...

@douglas-reid I have created https://github.com/istio/istio/issues/11919 for a broader discussion of the issue.  It should be simple to integrate a fix with the fix for istio#10134 as you mentioned.

**Comment by douglas-reid** [comment]
@mandarjog those requirements are pretty broad. can you provide comment on how this PR addresses your requirements specifically? I'm really not looking for a treatise, but just some clearly delineated rationale and explanation for the choices. The PR involves more than just a modification to add PD to prometheus, which is why I'm asking.

**Comment by therealmitchconnors** [comment]
@douglas-reid the core problem this PR intends to solve is the writing of prometheus data to an ephemeral volume.  The solution to this problem is to use a persistent volume (I did not consider remote-write or federation because they require resources outside the cluster, and we want to be able to spin up new clusters ad-hoc).  

However, a persistent volume claim cannot be attached to a deployment, which is how we are managing prometheus today.  Additionally, persistent volume semantics differ from one cloud to another, which means adding this behavior to the istio chart will be complicated.

Therefore the goal became to modify prometheus to use a persistent volume without modifying the istio chart.  This required disabling the deployment of prometheus from within the istio chart altogether, and deploying something with a persistent volume claim outside the chart.  This could have been either a StatefulSet or a Prometheus-Operator chart.  Prometheus-Operator is preferable here because it promotes code reuse, and gets us out of the business of maintaining prometheus best practices.

The additional trappings that come with the operator (kube-state-metrics, etc.) are incidental.  While they seem like valuable metrics to have, I am happy to remove them to focus on the primary issue of metrics retention.

**Comment by douglas-reid** [comment]
@therealmitchconnors thanks for your reply. how does the need for prometheus differ from the need for grafana which was addressed in https://github.com/istio/istio/pull/7553 ?  in other words, why does adding a PVC with some helm options not work? I'm curious here, as I have no real experience with PVs.

**Comment by therealmitchconnors** [comment]
@douglas-reid on further research, it looks like PVCs can in fact be used outside of statefulsets (this is my first time working with them as well).  I can't find the blog post which had indicated otherwise, but the K8S docs do show PVCs being used directly by Pods, and the PR you referenced seems to demonstrate them being used in deployments as well.

With this new information in mind, it may be possible to add similar flags for prometheus in the istio chart, and use those flags from within the test to use a persistent volume for prometheus.  Would that be your preference?  I am a little worried about changing something like this so close to the 1.1 release...

**Comment by douglas-reid** [comment]
@therealmitchconnors is there a way to add it only to the perf install, via the scripts (approach similar to this PR)?

I'm not all that worried about adding install options around persistence to 1.1 if they are needed for testing, but maybe doing it special in the perf scripting is the escape hatch.

**Comment by jnativio** [comment]
@mandarjog  That is what I did on the regression patrol: I have a script that modifies the Prometheus deployment in istio.yaml:

in container args, set the retention period and the db path:
            - '--storage.tsdb.retention=30d'
            - '--storage.tsdb.path=/prom_store/data'

in volumeMounts, mount a new volume:
          - name: myvol
            mountPath: /prom_store

and in the volumes, define the volume and run a initContainer to give the right permission: 
      - name: myvol
        persistentVolumeClaim:
          claimName: mypvc-bronze2
      initContainers:
      - name: permissionsfix
        image: alpine:latest
        command: ["/bin/sh", "-c"]
        args:
          - chown 65534:65534 /prom_test;
        volumeMounts:
        - name: myvol
          mountPath: /prom_test

Of course, I have already the PVC defined. It is a good idea to have Prometheus and the PVC in another namespace as I have deleted accidentally it a couple of times when uninstalling Istio.

**Comment by therealmitchconnors** [comment]
@jnativio can you point us to this script for reference?  

@douglas-reid I am usually hesitant to write a script to modify another script, but if that is working for us elsewhere, I can try to follow that pattern.

**Comment by therealmitchconnors** [comment]
@douglas-reid I have a little more context around PVCs and Deployments.  VolumeClaimTemplates are only allowed on Stateful Sets, meaning that deployments using persistent volumes have to declare each volume separately.  This means that if the deployment ever scales up, the same volume will be mounted to all instances, which would cause write contention in this case.  It seems unlikely that we will scale up prometheus (I don't see an HPA for that in the chart), but it is something to consider...

**Comment by jnativio** [comment]
@therealmitchconnors I use sed to modify the Prometheus deployment. It is not pretty, but works:
# changes for Prometheus
new_block1="\         - name: myvol\n\            mountPath: \/prom_store\n\      nodeSelector:\n\        nodename: worker4"
new_block2="\     - name: myvol\n        persistentVolumeClaim:\n\          claimName: mypvc-bronze2\n\      initContainers:\n\      - name: permissionsfix\n\        image: alpine:latest\n\        command: [\"\/bin\/sh\", \"-c\"]\n\        args:\n\          - chown 65534:65534 \/prom_test;\n\        volumeMounts:\n\        - name: myvol\n\          mountPath: \/prom_test"

sed -i '/mountPath: \/etc\/prometheus/{N;N;s/$/\n '"${new_block1}"'/}' $install_path/istio-helm_placed.yml
sed -i '/mountPath: \/etc\/prometheus/{N;N;N;N;N;N;N;N;N;N;N;N;N;N;N;s/$/\n '"${new_block2}"'/}' $install_path/istio-helm_placed.yml

sed -ie 's/'"- '--storage.tsdb.retention=6h'"'/'"- '--storage.tsdb.retention=30d'\n            - '--storage.tsdb.path=\/prom_store\/data'"'/g' $install_path/istio-helm_placed.yml

**Comment by therealmitchconnors** [comment]
I've spent some time over the last few days trying to get a chart put together that would mount a persistent volume to the existing prometheus chart.  While it is not impossible, it requires substantial changes to get prometheus running as a user with access to the persistent volume, a feat which is accomplished out of the box with the prometheus operator.  I recommend we adopt the operator in our qualification tests to evaluate its effectiveness and ease of use, with a view towards leveraging it in the helm chart, or future istio install workflows.

**Comment by douglas-reid** [comment]
@therealmitchconnors ok. thanks for effort. i suspect that will not be in vain.

**Comment by mandarjog** [comment]
@therealmitchconnors can you resolve conflict. That file was renames.

**Comment by mandarjog** [comment]
Hey Mitch files have moved around. 

Please create a directory `perf/prometheus-install`  and move all your scripts and configuration there.

File: perf/prometheus-install/prometheus-scrape-configs.yaml [hunk 1]
```diff
@@ -0,0 +1,286 @@
+
+- job_name: 'istio-mesh'
+  kubernetes_sd_configs:
+  - role: endpoints
+    namespaces:
+      names:
+      - istio-system
+
```
Comment (mandarjog) [inline perf/prometheus-install/prometheus-scrape-configs.yaml]: Is this same or different than the stock prom config?

Comment (therealmitchconnors) [reply perf/prometheus-install/prometheus-scrape-configs.yaml]: If by stock, you mean the istio chart, this is the same namespace, although the service name does change from just prometheus to istio-prometheus (this is unavoidable).

File: perf/istio-install/setup_istio.sh [hunk 2]
```diff
@@ -33,6 +33,33 @@ function setup_admin_binding() {
...
+  # Store original context namespace so it can be reset at the end
+  local ORIG_CTX=$(kubectl config current-context)
+  local ORIG_NS=$(kubectl config get-contexts ${ORIG_CTX} --no-headers | tr -s ' ' | cut -d ' ' -f 5)
+  # Prometheus operator chart doesn't respect --namespace, all objects are
+  # deployed to the default namespace.
+  kubectl config set-context $(kubectl config current-context) --namespace=istio-system
+  helm template "${PROMETHEUS_INSTALL}/prometheus-operator/"\
+    -f "${PROMETHEUS_INSTALL}/prometheus-operator-values.yaml"\
+    --set-file .Values.prometheus.prometheusSpec.additionalScrapeConfigs="${PROMETHEUS_INSTALL}/prometheus-scrape-configs.yaml"\
+    | kubectl apply -f -
```
Comment (howardjohn) [inline perf/istio-install/setup_istio.sh]: `kubectl apply -n istio-system` should work without all the context hacks?

File: perf/prometheus-install/prometheus-operator-values.yaml [hunk 3]
```diff
@@ -0,0 +1,1103 @@
+# Values for prometheus-operator.
```
Comment (howardjohn) [inline perf/prometheus-install/prometheus-operator-values.yaml]: Was there a reason you copied all of this? You can just make a file with just the changes you apply, no need to have the defaults copied - it makes it impossible to tell what you changed

Comment (mandarjog) [reply perf/prometheus-install/prometheus-operator-values.yaml]: Good point. Please make this change along with updating the the istio base prom config to be aware of operator in a subsequent PR.

File: perf/istio-install/setup_istio.sh (hunk 2 above)
Comment (therealmitchconnors) [reply perf/istio-install/setup_istio.sh]: I tried that, but had difficulty... it was almost a month ago now, and I can't recall what the problem was.  I will test it out and report back.

File: perf/prometheus-install/prometheus-operator-values.yaml (hunk 3 above)
Comment (therealmitchconnors) [reply perf/prometheus-install/prometheus-operator-values.yaml]: > /lgtm
> 
> We should templatize the istio standard prometheus config to support this so we don't include a special config here.

I agree that this should be the long term direction, but I am not yet confident enough of the operator to push it out so broadly.  My intent was to prove the operator here in the release qual testing before deploying it more fully, along with the install work @costinm is doing...

**Pull request merged**